File: kolko2/r_22-23/python/ogrodzenie/main.py
from math import sqrt, atan2, pi


def orientation(p1, p2, p3):
    return ((p2[1] - p1[1]) * (p3[0] - p2[0]) - (p2[0] - p1[0]) * (p3[1] - p2[1])) >= 0


def grahamScan(points, n):
    border = [0, 1]
    p = 0
    c = 1
    for next in range(2, n):
        if orientation(points[p], points[c], points[next]):
            border.pop()
        border.append(next)
        c = border[-1]
        p = border[-2]
    border.append(0)
    return sum(
        [
            sqrt(
                (points[border[i]][0] - points[border[i - 1]][0]) ** 2
                + (points[border[i]][1] - points[border[i - 1]][1]) ** 2
            )
            for i in range(1, len(border))
        ]
    )


def main():
    n = int(input())
    points = sorted(
        [list(map(int, input().split())) for _ in range(n)], key=lambda p: (p[1], p[0])
    )
    points = [points[0]] + sorted(
        points[1:],
        key=lambda p: (
            atan2(p[1] - points[0][1], p[0] - points[0][0]) * 180 / pi,
            sqrt((p[0] - points[0][0]) ** 2 + (p[1] - points[0][1]) ** 2),
        ),
    )
    borderLen = grahamScan(points, n)
    print("0.%10g" % borderLen)


if __name__ == "__main__":
    main()

File: kolko2/r_22-23/cpp/ogrodzenie/main.cpp
#include <algorithm>
#include <cmath>
#include <iomanip>
#include <iostream>
#include <vector>

using namespace std;

struct point {
    int x, y;
};

const int MAXN = 5e5 + 7;
int lowest, n, i;
long double borderLen;
vector<int> border;
point points[MAXN];

bool orientation(point p1, point p2, point p3) {
    return ((p2.y - p1.y) * (p3.x - p2.x) - (p2.x - p1.x) * (p3.y - p2.y)) >= 0;
}

bool porownajKat(point a, point b) {
    long double angleA =
        atan2(a.y - points[0].y, a.x - points[0].x) * 180 / M_PI;
    long double angleB =
        atan2(b.y - points[0].y, b.x - points[0].x) * 180 / M_PI;
    if (angleA == angleB) {
        long double distA =
            sqrt(pow(a.x - points[0].x, 2) + pow(a.y - points[0].y, 2));
        long double distB =
            sqrt(pow(b.x - points[0].x, 2) + pow(b.y - points[0].y, 2));
        return distA < distB;
    }
    return angleA < angleB;
}

void countBorderLen() {
    int prev = 0, curr = 1;
    border.push_back(0);
    border.push_back(1);
    for (int next = 2; next < n; ++next) {
        if (orientation(points[prev], points[curr], points[next])) {
            border.pop_back();
        }
        prev = border[border.size() - 1];
        border.push_back(next);
        curr = border[border.size() - 1];
    }
    border.push_back(0);
    for (i = 0; i < border.size(); ++i) {
        borderLen +=
            sqrt(pow((points[border[i]].x - points[border[i - 1]].x), 2) +
                 pow((points[border[i]].y - points[border[i - 1]].y), 2));
    }
}

int main() {
    ios_base::sync_with_stdio(0);
    cin.tie(0);
    cin >> n;
    for (i = 0; i < n; ++i) {
        cin >> points[i].x >> points[i].y;
        if ((points[i].y < points[lowest].y) ||
            ((points[i].y == points[lowest].y) &&
             (points[i].x < points[lowest].x)))
            lowest = i;
    }
    swap(points[lowest], points[0]);
    sort(points + 1, points + n, porownajKat);
    countBorderLen();
    cout << fixed << setprecision(10) << borderLen << "\n";
}